src/trackinfoobject.cpp

```diff
@@ -861,64 +832,61 @@ void TrackInfoObject::slotCueUpdated() {
     emit(cuesUpdated());
 }
 
-Cue* TrackInfoObject::addCue() {
+CuePointer TrackInfoObject::addCue() {
     //qDebug() << "TrackInfoObject::addCue()";
     QMutexLocker lock(&m_qMutex);
-    Cue* cue = new Cue(m_iId);
-    connect(cue, SIGNAL(updated()),
+    CuePointer pCue(new Cue(m_iId));
+    connect(pCue.data(), SIGNAL(updated()),
             this, SLOT(slotCueUpdated()));
-    m_cuePoints.push_back(cue);
-    setDirty(true);
-    lock.unlock();
+    m_cuePoints.push_back(pCue);
+    setDirtyAndUnlock(&lock, true);
     emit(cuesUpdated());
-    return cue;
+    return pCue;
 }
 
-void TrackInfoObject::removeCue(Cue* cue) {
+void TrackInfoObject::removeCue(CuePointer pCue) {
     QMutexLocker lock(&m_qMutex);
-    disconnect(cue, 0, this, 0);
-    // TODO(XXX): Delete the cue point.
-    m_cuePoints.removeOne(cue);
-    setDirty(true);
-    lock.unlock();
+    disconnect(pCue.data(), 0, this, 0);
+    m_cuePoints.removeOne(pCue);
+    setDirtyAndUnlock(&lock, true);
     emit(cuesUpdated());
 }
 
-const QList<Cue*>& TrackInfoObject::getCuePoints() {
+QList<CuePointer> TrackInfoObject::getCuePoints() const {
     QMutexLocker lock(&m_qMutex);
     return m_cuePoints;
 }
 
-void TrackInfoObject::setCuePoints(QList<Cue*> cuePoints) {
+void TrackInfoObject::setCuePoints(QList<CuePointer> /*by-value*/ cuePoints) {
```

Comment:
@rryan Fixed, parameter passed by const-ref + using assignment operator

Implicit sharing is a questionable concept of Qt that has it's roots in the pre C++0x era deep in the 90s. But even the first STL implementations dropped their COW versions of std::string for a reason. With C++11 this concept finally becomes obsolete or even counter productive.

Simple example:
void doSomething(QStringList someStrings) {
for (int i = 0; i < someStrings.size(); ++i) {
    const auto& s = someStrings[i];
    // The compiler will invoke non-const operator[] even if we
    // only need a const-ref here. The left side of the assignment
    // operator is not taken into account when selecting an overloaded
    // member function. The compiler always favours the non-const
    // over the const variant unless 'this' itself is const!
    ...
}

The first invocation of non-const operator[] needs to detach the local instance from the shared data and creates a deep-copy! The situation might improve when using Qt's foreach macro with a "const QString&" argument (I haven't analysed it), but the main critic remains. If you don't use const rigorously throughout your code you will get unintended copy operations where you don't expect them. You simply don't see what is actually happening behind the scenes.

Qt is handy for the GUI layer and even inexperienced C++ developers are able to use it. But when you need to exploit the real power of C++ I strongly recommend to use the standard containers from the STL.